src/frontend/src/optimizer/rule/apply_join.rs

```diff
@@ -0,0 +1,302 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use fixedbitset::FixedBitSet;
+use itertools::Itertools;
+use risingwave_common::types::DataType;
+use risingwave_common::types::DataType::Boolean;
+use risingwave_pb::plan_common::JoinType;
+
+use super::{BoxedRule, Rule};
+use crate::expr::{
+    CollectInputRef, CorrelatedId, CorrelatedInputRef, Expr, ExprImpl, ExprRewriter, ExprType,
+    ExprVisitor, FunctionCall, InputRef,
+};
+use crate::optimizer::plan_correlated_id_finder::PlanCorrelatedIdFinder;
+use crate::optimizer::plan_node::{
+    LogicalApply, LogicalFilter, LogicalJoin, LogicalProject, PlanTreeNodeBinary,
+};
+use crate::optimizer::plan_visitor::PlanVisitor;
+use crate::optimizer::PlanRef;
+use crate::utils::{ColIndexMapping, Condition};
+
+/// Push `LogicalJoin` down `LogicalApply`
+/// D Apply (T1 join<p> T2)  ->  (D Apply T1) join<p and natural join D> (D Apply T2)
```

Comment:
(Not a comment. Just want to attach a [digram](https://dreampuf.github.io/GraphvizOnline/#digraph%20G%20%7B%0A%0A%20%20subgraph%20cluster_0%20%7B%0A%20%20%20%20Apply0%20-%3E%20D0%0A%20%20%20%20Apply0%20-%3E%20Join0%0A%20%20%20%20Join0%20-%3E%20L0%0A%20%20%20%20Join0%20-%3E%20R0%0A%20%20%7D%0A%0A%20%20subgraph%20cluster_1%20%7B%0A%20%20%20%20Join1%20-%3E%20Apply1l%0A%20%20%20%20Join1%20-%3E%20Apply1r%0A%20%20%20%20Apply1l%20-%3E%20D1l%0A%20%20%20%20Apply1l%20-%3E%20L1%0A%20%20%20%20Apply1r%20-%3E%20D1r%0A%20%20%20%20Apply1r%20-%3E%20R1%0A%20%20%7D%0A%0A%7D) here.)